modules/02_Theory.md

```diff
@@ -0,0 +1,371 @@
+
+> There are [many module systems](https://github.com/myshov/history-of-javascript/tree/master/4_evolution_of_js_modularity), and TypeScript [supports emitting several](https://www.typescriptlang.org/tsconfig/#module), but this documentation will focus on the two most important systems today: ECMAScript modules (ESM) and CommonJS (CJS).
+>
+> ECMAScript Modules (ESM) is the module system built into the language, supported in modern browsers and in Node since v12. It uses dedicated `import` and `export` syntax:
+>
+> ```js
+> // a.js
+> export default "Hello from a.js";
+> ```
+>
+> ```js
+> // b.js
+> import a from "./a.js";
+> console.log(a); // 'Hello from a.js'
+> ```
+>
+> CommonJS (CJS) is the module system that originally shipped in Node, before ESM was part of the language specification. It’s still supported in Node alongside ESM. It uses plain JavaScript objects and functions named `exports` and `require`:
+>
+> ```js
+> // a.js
+> exports.message = "Hello from a.js";
+> ```
+>
+> ```js
+> // b.js
+> const a = require("./a");
+> console.log(a.message); // 'Hello from a.js'
+> ```
+
+Accordingly, when TypeScript detects that a file is a CommonJS or ECMAScript module, it starts by assuming that file will have its own scope. Beyond that, though, the compiler’s job gets a little more complicated.
+
+## TypeScript’s job concerning modules
+
+The TypeScript compiler’s chief goal is to look at input code and tell the author about problems the output code might encounter at runtime. To do this, the compiler needs to know some things about the code’s intended runtime environment—what globals are available, for example. When the input code is a module, there are several additional questions the compiler needs to answer in order to do its job. Let’s use a few lines of input code as an example to think about all the information needed to analyze it:
+
+```ts
+import sayHello from "greetings";
+sayHello("world");
+```
+
+To check this file, the compiler needs to know the type of `sayHello` (is it a function that can accept one string argument?), which opens quite a few additional questions:
+
+1. What kind of module does the module system expect to find at the location where I output this file?
+2. What output JavaScript code should I emit for this input?
+3. When the module system loads that output code, where will it look to find the module specified by `"greetings"`? Will the lookup succeed?
+4. What kind of module is the file resolved by that lookup?
+5. Does the module system allow the kind of module detected in (1) to reference the kind of module detected in (4) with the syntax decided in (2)?
+6. Once the `"greetings"` module has been analyzed, what piece of that module is bound to `sayHello`?
+
```

Comment:
I think at this point we want to reflect on the runtime execution of this code, specifically as relates to the resolution of module specifiers as an abstract operation. This gives a good to define "host" and "module resolution" in an agnostic way and lets the following paragraph have fewer embedded definitions.